toolchain/check/context.cpp

```diff
@@ -1233,13 +1233,22 @@ class TypeCompleter {
 };
 }  // namespace
 
-auto Context::TryToCompleteType(SemIR::TypeId type_id,
+auto Context::TryToCompleteType(SemIR::TypeId type_id, SemIR::LocId loc_id,
                                 BuildDiagnosticFn diagnoser,
                                 BuildDiagnosticFn abstract_diagnoser) -> bool {
   if (!TypeCompleter(*this, diagnoser).Complete(type_id)) {
     return false;
   }
 
+  if (diagnoser && type_id.AsConstantId().is_symbolic()) {
```

Comment:
I'm trying to understand exactly what's going on here. Looking at the comments on the declaration, I _think_ it's something like (and a comment would help confirm):

```suggestion
  // Resolve dependent types when a complete type is required.
  if (diagnoser && type_id.AsConstantId().is_symbolic()) {
```

But, had you considered adding overloads to split out logic instead of using default parameters? Looking at calls, it seems like `abstract_diagnoser` requires `diagnoser. So I'm wondering if this would make sense as a split:

```
TryToCompleteType(type_id, loc_id) -> bool {
  return TypeCompleter(*this, /*diagnoser=*/nullptr).Complete(type_id);
}
RequireCompleteType(type_id, loc_id, diagnoser) -> bool {
  CARBON_CHECK(diagnoser);
  // current function up to abstract_diagnoser
}
RequireConcreteType(type_id, loc_id, diagnoser, abstract_diagnoser) {
  CARBON_CHECK(diagnoser && abstract_diagnoser);
  if (!RequireCompleteType(type_id, loc_id, diagnoser)) {
    return false;
  }
  // abstract_diagnoser logic
}
```

That might also simplify how comments are written, since diagnoser functions could be required.

Reply:
Added a comment.

For the refactoring, I agree; I had been planning to do something like that as a follow-up. Can do it in this PR if you prefer.

Reply:
It's probably going to touch a bunch of callers, so fine to split it out.

Reply:
Well, I guess not that many callers. But still fine to split it out.

Reply:
#4658